templates/commands/render.go

```diff
@@ -235,24 +275,237 @@ func (r *Render) realRun(ctx context.Context, rp *runParams) (outErr error) {
 		return err
 	}
 
-	_ = templateDir // TODO: add template rendering logic
+	spec, err := loadSpecFile(rp.fs, templateDir, r.flagSpec)
+	if err != nil {
+		return err
+	}
+
+	scratchDir, err := rp.tempDirNamer("scratch")
+	if err != nil {
+		return err
+	}
+	if err := rp.fs.MkdirAll(scratchDir, 0o700); err != nil {
+		return fmt.Errorf("MkdirAll(): %w", err)
+	}
 
+	if err := executeSpec(ctx, spec, &stepParams{
+		inputs:      r.flagInputs,
+		fs:          rp.fs,
+		scratchDir:  scratchDir,
+		stdout:      rp.stdout,
+		templateDir: templateDir,
+	}); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func executeSpec(ctx context.Context, spec *model.Spec, sp *stepParams) error {
+	for _, step := range spec.Steps {
+		if err := executeOneStep(ctx, step, sp); err != nil {
+			return err
+		}
+	}
 	return nil
 }
 
+type stepParams struct {
+	fs          renderFS
+	inputs      map[string]string
+	scratchDir  string
+	stdout      io.Writer
+	templateDir string
+}
+
+func executeOneStep(ctx context.Context, step *model.Step, sp *stepParams) error {
+	switch {
+	case step.Print != nil:
+		return actionPrint(ctx, step.Print, sp)
+	case step.Include != nil:
+		return actionInclude(ctx, step.Include, sp)
+	case step.RegexReplace != nil:
+		return actionRegexReplace(ctx, step.RegexReplace, sp)
+	case step.StringReplace != nil:
+		return actionStringReplace(ctx, step.StringReplace, sp)
+	case step.GoTemplate != nil:
+		return actionGoTemplate(ctx, step.GoTemplate, sp)
+	default:
+		return fmt.Errorf("internal error: unknown step action type %q", step.Action.Val)
+	}
+}
+
+func errWithPos(pos *model.ConfigPos, fmtStr string, args ...any) error {
+	err := fmt.Errorf(fmtStr, args...)
+	lineStr := fmt.Sprintf("line %d", pos.Line)
+	if pos.Line == 0 {
+		lineStr = "(unknown line number)"
+	}
+	return fmt.Errorf("failed executing template spec file at %s: %w", lineStr, err)
+}
+
+func parseGoTmpl(tpl string) (*template.Template, error) {
+	return template.New("").Option("missingkey=error").Parse(tpl)
+}
+
+func parseAndExecuteGoTmpl(m model.String, inputs map[string]string) (string, error) {
+	goTmpl, err := parseGoTmpl(m.Val)
```

Comment:
Does it make sense to cache any of these templates so we don't keep reparsing?

Reply:
That will come soon. In the current version of the code, there's very little template reuse. Once we add the regex_replace action we'll want to cache both parsed templates and parsed regexes, since they'll be reused in a loop.

This exists in my prototype, it's just not time yet for this code.